File: src/AbOvo.Contracts/Rates.cs
using System.ComponentModel.DataAnnotations;

namespace AbOvo.Contracts;

/// <summary>
/// A proportion and the interval around it, as ONE value.
///
/// <para>
/// Issue #16's whole requirement: <em>"Not two fields where the second is optional: an
/// optional interval is an interval that will be absent on the screen where it matters, and
/// a bare rate is exactly the 'ratio quoted without its two quantities' the book spends a
/// program complaining about."</em>
/// </para>
/// <para>
/// So there is no public constructor and no settable property. <see cref="Of"/> is the only
/// way to obtain one, it computes every field from the two counts, and it refuses a total of
/// zero — see its own note for why that is an absence rather than a rate of zero.
/// </para>
/// <para>
/// <b>It follows that a Rate cannot be DESERIALISED, and that is deliberate.</b> The obvious
/// fix for the build error that produces — <c>[JsonConstructor]</c> on the constructor below
/// — compiles and silently undoes the whole point: a deserialiser fills absent fields with
/// their default, so a payload carrying only <c>passed</c>, <c>total</c> and <c>percent</c>
/// would arrive as a rate whose interval is zero points wide. A FAKE interval is worse than a
/// missing one, because it renders, it sorts, and it says there is no uncertainty. This type
/// is produce-only; <c>RatesCarryTheirIntervalTests.A_rate_cannot_be_deserialised</c> pins it.
/// </para>
/// <para>
/// <b>WHAT THE INTERVAL DOES NOT ACCOUNT FOR</b>, and it is not a rounding detail. The
/// arithmetic assumes the items are independent. Whether they are is a property of the CELL
/// this rate was computed over, not of this type — Program P27 measures the cost of getting
/// that wrong at a factor of <c>sqrt(rows per reader)</c>, and this service has no reader
/// identifier and so cannot measure it. `Proportion` in the API states which cell is safe
/// and why; ADR-0024 records the decision.
/// </para>
/// </summary>
public sealed record Rate
{
    private Rate(long passed, long total, double percent, double halfWidth, double low, double high)
    {
        Passed = passed;
        Total = total;
        Percent = percent;
        HalfWidth = halfWidth;
        Low = low;
        High = high;
    }

    /// <summary>How many of <see cref="Total"/> passed.</summary>
    public long Passed { get; }

    /// <summary>The denominator. Always at least 1 — see <see cref="Of"/>.</summary>
    public long Total { get; }

    /// <summary>The rate, in percentage points.</summary>
    public double Percent { get; }

    /// <summary>
    /// Half the interval's width, in percentage points.
    ///
    /// <para>
    /// Carried as well as the two endpoints, deliberately: it is the quantity a reader of the
    /// screen compares between rows, and deriving it there would be the same arithmetic in a
    /// second place. <see cref="Low"/> and <see cref="High"/> are clamped to [0, 100] and this
    /// is not, so <c>High - Low</c> is NOT twice this near either end — which is the property
    /// that would make a derived version wrong rather than merely duplicated.
    /// </para>
    /// </summary>
    public double HalfWidth { get; }

    /// <summary>The interval's lower end, clamped to 0.</summary>
    public double Low { get; }

    /// <summary>The interval's upper end, clamped to 100.</summary>
    public double High { get; }

    /// <summary>
    /// The only way to make one.
    ///
    /// <para>
    /// <b>A total of zero is refused rather than reported as 0%.</b> With no observations
    /// there is no proportion: <c>p</c> is undefined, and the honest rendering of "nobody has
    /// run this check" is an absent measurement, not a measured zero. A rate of 0% with a wide
    /// interval reads as <em>every reader failed</em>, which is the worst available misreading
    /// and the one a ranked list puts at the top. The caller omits the cell instead.
    /// </para>
    /// </summary>
    /// <param name="passed">How many passed. Must not exceed <paramref name="total"/>.</param>
    /// <param name="total">How many ran. Must be at least 1.</param>
    /// <param name="halfWidth">
    /// The interval's half-width in percentage points, from the API's <c>Proportion</c>. Passed
    /// in rather than computed here because <c>AbOvo.Contracts</c> is the shape on the wire and
    /// holds no arithmetic — and because the choice of <c>z</c> is a decision with an owner
    /// (Program P27) rather than a constant this record should carry.
    /// </param>
    public static Rate Of(long passed, long total, double halfWidth)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(total, 1);
        ArgumentOutOfRangeException.ThrowIfNegative(passed);
        ArgumentOutOfRangeException.ThrowIfGreaterThan(passed, total);
        ArgumentOutOfRangeException.ThrowIfNegative(halfWidth);

        var percent = 100.0 * passed / total;
        return new Rate(
            passed,
            total,
            percent,
            halfWidth,
            Math.Max(0.0, percent - halfWidth),
            Math.Min(100.0, percent + halfWidth));
    }
}

/// <summary>
/// One cell's rate: a check, at an attempt, on a frame.
///
/// <para>
/// <b>THE CELL IS THE UNIT AND IT IS NOT A CHOICE OF GRANULARITY.</b> It is the finest cell
/// the store has, and it is also the only one over which <see cref="Rate"/>'s arithmetic is
/// sound — pooling across checks or across attempts pools observations that share a reader,
/// and an interval computed over those is too narrow by a factor nothing here can measure.
/// ADR-0024 §2.
/// </para>
/// </summary>
public sealed record CellRate
{
    [Range(1, int.MaxValue)]
    public int Step { get; init; }

    [Required]
    public string Check { get; init; } = string.Empty;

    [Range(1, 50)]
    public int Attempt { get; init; }

    [Required]
    public required Rate Rate { get; init; }
}

/// <summary>
/// Every measured cell of one unit, at one bundle tag.
///
/// <para>
/// The tag is on the envelope rather than on each cell because it is the same for all of
/// them BY CONSTRUCTION: the query pins it, and a response mixing two tags would be the
/// average across two texts that <c>FrameOutcome</c>'s key exists to prevent. Repeating it
/// per row would invite a caller to believe rows could differ.
/// </para>
/// </summary>
public sealed record UnitRates
{
    [Required]
    public string BundleTag { get; init; } = string.Empty;

    [Required]
    public string Track { get; init; } = string.Empty;

    [Required]
    public string Unit { get; init; } = string.Empty;

    /// <summary>
    /// The cells that have at least one observation, in <c>(step, check, attempt)</c> order.
    ///
    /// <para>
    /// A unit nobody has run yet is an empty array and a 200, not a 404: the unit exists, and
    /// "no evidence yet" is an answer rather than an absence. It is also the state every unit
    /// starts in, so a 404 there would make the instrument's first week look like a bug.
    /// </para>
    /// </summary>
    public IReadOnlyList<CellRate> Cells { get; init; } = [];
}

File: web/app/src/lib/instrument/rates.ts
/**
 * The instrument's rates, on this side of the wire.
 *
 * ──────────────────────────────────────────────────────────────────────────────────────
 * A TYPE IS NOT A GUARANTEE, AND ISSUE #16 ASKS FOR A GUARANTEE.
 *
 * "The rate and its interval are one non-nullable value, on the C# record and on the
 * generated TypeScript type." A TypeScript interface is erased at run time, so
 * `JSON.parse(body) as UnitRates` promises exactly nothing: a payload that lost `halfWidth`
 * in transit, or came from an older build, or came from something else entirely, arrives as
 * an object whose `rate.halfWidth` is `undefined` and renders as `NaN` — or, worse, as a
 * blank cell beside a rate that then reads as certain.
 *
 * So the type is paired with `readRates`, which refuses. Every field is required, every
 * number must be finite, and a payload missing one is not a partially-good response — it is
 * not a response. That is P11 at the edge, and it is the same direction `parse.ts` takes one
 * module over: **a wrong number is worse than no number**, because a wrong one renders.
 * ──────────────────────────────────────────────────────────────────────────────────────
 *
 * THE SHAPE HAS ONE SOURCE AND IT IS NOT THIS FILE. `src/AbOvo.Contracts/rates.contract.json`
 * is a serialised `UnitRates` from the real records, committed; the C# suite asserts the
 * records produce exactly it, and `rates.test.ts` asserts this module consumes exactly it.
 * Neither side can add, rename or drop a field alone. There is no code generator in this
 * repository and this is what stands in for one — see the PR for issue #16, which says so
 * rather than letting "generated" be read into it.
 */

/** A proportion and its interval, which arrive together or not at all. */
export interface Rate {
  readonly passed: number;
  readonly total: number;
  /** The rate, in percentage points. */
  readonly percent: number;
  /**
   * Half the interval's width, in percentage points, UNCLAMPED.
   *
   * `high - low` is not twice this near either end, because the ends are clamped to [0, 100]
   * and this is not. That is deliberate: this is the quantity that says how much evidence
   * there is, and clamping it would make two cells with very different evidence look alike
   * near the ceiling.
   */
  readonly halfWidth: number;
  readonly low: number;
  readonly high: number;
}

/** One cell: a check, at an attempt, on a frame. Never a pooled rate — see the API's `Proportion`. */
export interface CellRate {
  readonly step: number;
  readonly check: string;
  readonly attempt: number;
  readonly rate: Rate;
}

/**
 * How much the extreme of a RANKED list overstates, by selection alone.
 *
 * Sort noisy estimates and read the end of it, and the end sits beyond the truth even when
 * every item is equally good — because sorting selects for whichever estimate the noise
 * pushed furthest. It is a property of the LIST, never of a cell: a row in the middle of a
 * ranking was not selected for, so this number beside every row would be a wrong number that
 * renders.
 *
 * COMPUTED BY THE API, never here. Program P27's arithmetic has one implementation and it is
 * the one `NormalGatesTests` gates against the book's own committed figures; a second copy on
 * this side would be a routine that has to agree with it and that nothing checks.
 */
export interface SelectionMargin {
  /** How many cells the ranking sorts through. At least one. */
  readonly ranked: number;
  /** `E[max of ranked standard normals]` — the margin in standard errors. */
  readonly standardErrors: number;
  /** The same margin in percentage points, at the extreme cell's own standard error. */
  readonly points: number;
}

export interface UnitRates {
  readonly bundleTag: string;
  readonly track: string;
  readonly unit: string;
  /** Only cells with at least one observation. A unit nobody has run is an empty array. */
  readonly cells: readonly CellRate[];
  /**
   * Absent when there are no cells, and present — as zero — when there is exactly one.
   *
   * Not `| undefined` by oversight: there is no list to select from, so "this ranking
   * overstates by nothing" is a sentence about something that does not exist, and it would
   * render beside an empty table as *this ranking is trustworthy*. One cell is different and
   * is reported, because selecting the extreme of one thing selects for nothing and that is a
   * fact rather than a placeholder.
   */
  readonly selection: SelectionMargin | null;
}

const isObject = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null && !Array.isArray(value);

/**
 * A finite number, and `Number.isFinite` rather than `typeof === 'number'`.
 *
 * `NaN` and `Infinity` are both `number`, both survive `JSON.parse` of a payload built by a
 * language that has them, and both render. A `NaN` half-width beside a rate is a cell that
 * says nothing where it looks like it says something.
 */
const num = (value: unknown): value is number =>
  typeof value === 'number' && Number.isFinite(value);

const text = (value: unknown): value is string => typeof value === 'string';

function readRate(value: unknown): Rate | null {
  if (!isObject(value)) return null;

  const { passed, total, percent, halfWidth, low, high } = value;
  if (!num(passed) || !num(total) || !num(percent)) return null;
  if (!num(halfWidth) || !num(low) || !num(high)) return null;

  return { passed, total, percent, halfWidth, low, high };
}

function readCell(value: unknown): CellRate | null {
  if (!isObject(value)) return null;

  const { step, check, attempt } = value;
  if (!num(step) || !text(check) || !num(attempt)) return null;

  const rate = readRate(value['rate']);
  // The whole point, in one line: a cell whose rate did not parse is not a cell with a
  // missing interval, it is not a cell.
  return rate === null ? null : { step, check, attempt, rate };
}

function readSelection(value: unknown): SelectionMargin | null {
  if (!isObject(value)) return null;

  const { ranked, standardErrors, points } = value;
  if (!num(ranked) || !num(standardErrors) || !num(points)) return null;

  // The properties the number means, refused rather than rendered. `ranked` under one is a
  // ranking of nothing; a negative margin would say that sorting a list makes its extreme
  // look BETTER than the truth, which is the opposite of what selection does and would render
  // as a correction pointing the wrong way.
  if (ranked < 1 || standardErrors < 0 || points < 0) return null;

  return { ranked, standardErrors, points };
}

/**
 * Parse an API response into rates, or `null`.
 *
 * `null` for the WHOLE document rather than the cells that happened to parse: a partially
 * read ranking is a ranking with rows silently missing, which reads as a shorter list rather
 * than as a broken one. ADR-0014 records the same refusal for a content bundle, in the
 * book's own words — "the compiler REFUSES rather than degrades."
 */
export function readRates(value: unknown): UnitRates | null {
  if (!isObject(value)) return null;

  const { bundleTag, track, unit, cells } = value;
  if (!text(bundleTag) || !text(track) || !text(unit)) return null;
  if (!Array.isArray(cells)) return null;

  const read: CellRate[] = [];
  for (const cell of cells) {
    const parsed = readCell(cell);
    if (parsed === null) return null;
    read.push(parsed);
  }

  /*
   * THE MARGIN IS REQUIRED EXACTLY WHEN THERE IS A LIST, and both halves are refusals.
   *
   * Cells and no margin is a ranking whose cost of being ranked went missing in transit, and
   * it would render as a table with no caveat beside it — which is the one outcome issue #17
   * is written against. No cells and a margin is a number about a list that does not exist.
   */
  const raw = value['selection'];

  if (read.length === 0) {
    // Absent and an explicit `null` both say the same thing here and there is nothing to be
    // wrong about, so both are accepted. A margin that is PRESENT is refused: it would be a
    // correction for a ranking of no rows.
    return raw === null || raw === undefined
      ? { bundleTag, track, unit, cells: read, selection: null }
      : null;
  }

  const selection = readSelection(raw);
  return selection === null ? null : { bundleTag, track, unit, cells: read, selection };
}
